### ISSUE TITLE
How to reduce number of deprecation warnings?

### ISSUE THREAD

**Issue by leoarnold** [natritmeyer/site_prism#325]
We have 262 Capybara / RSpec examples. Using SitePrism 2.17 we get the warning

```
This config option is being removed in SitePrism v3. See UPGRADING.md on the repo for more details.
Going forwards the configuration / configurability will more closely mimic Capybara
```

4837 times (i.e. more than 18 times per example) in our build log, which is way more than the Travis CI website is capable to display.

I think there is at least one warning every time we expect existence of an element, i.e. we pass this line https://github.com/natritmeyer/site_prism/blob/v2.17/lib/site_prism/element_container.rb#L181

Is there any way to reduce the number of warnings?

**Comment by luke-hill** [comment]
Hi There Leo,

So ordinarily I'd agree with you (And I do agree that 4000+ is insane). Few points to note though.
- `3.0.beta` when released removes all deprecated code (This will be released in the next week)
- The config warnings you've mentioned there are only triggered when you either read or write to a config setting. So that means (If this is the only warning you get), that you read/write to a config setting on average 18 times per individual unit test? That sounds .... implausible?
- However notwithstanding the above, yes, we do read the setting for implicit waits when performing the `has_?` wait check.

Considering the above I am happy for any of the following...
- a PR which introduces a new setting to remove all warning messages, it itself would fire one warning message on initialisation (Don't like this personally*)
- a PR which reduces the number of warnings / which messages fire warnings by setting some class instance variables so they only fire a maximum of say 3 times? (Better)

*I am planning on releasing a SitePrism logger where a lot of warning / error messages will bubble up and not pollute the stdout side of things, but that's likely to take a while longer to implement.

Considering all of the above, if you're happy raising a PR then I'd happily accept it. Also it's worth considering all of the warnings will disappear in `3.0`, with a beta cut of the gem coming in the next few days - It might be worth trying to update?? There's an [UPGRADING.md](https://github.com/natritmeyer/site_prism/blob/master/UPGRADING.md) document here that briefly outlines the stuff you'd need to do / change.

I hope that explains / helps. Let me know if I can do anything else.

**Comment by leoarnold** [comment]
@luke-hill Thanks for providing some insight and outlining the roadmap.

Yes, that part of our code base is not the best, featuring way more than one expectation per example, lots `have_?` matchers there, hence frequently passing through

https://github.com/natritmeyer/site_prism/blob/eb68bd0cef120b281a12ba76d8117727b56fa84f/lib/site_prism/element_container.rb#L15-L21

I'd agree with you that waiting for 3.0 (or betas thereof) is probably best. Maybe we'll just keep this issue open nevertheless for others to find your explanations until 3.0 ships.

**Comment by mnohai-mdsol** [comment]
I too get the error, however, it occurs when calling ```set``` on an element. Example:

```ruby
require_relative '../common/base_page'

module TestApp
  class Login < Common::BasePage

    set_url $config['ui']['apps']['test_app']['base_url']

    element :username, 'input#session_username'
    element :password, 'input#session_password'
    element :submit, '#create_session_link'

    set_url_matcher(/login/)

    def initialize
      load
    end

    def login(user_name, pass)
      username.set user_name
      password.set pass
      submit.click
    end
  end
```

**Comment by leoarnold** [comment]
For the time being, I now added this monkey patch to our project

```ruby
# spec/spec_helper.rb
require 'site_prism/version'

fail "The monkey patch in #{__FILE__} is now obsolete. Please delete it." if Gem::Version.new(SitePrism::VERSION) >= Gem::Version.new('3')

module SitePrism
  class << self
    def show_removed_config_warning_messages; end
  end
end
```

**Comment by luke-hill** [comment]
Hi @leoarnold 

I've cut both `3.0.beta` and `2.17.1` this morning and both are released to rubygems.

For 2.17.1 I've put in a couple of bugfixes / optimisations, mainly I've reduced the warnings from appearing on all config settings, to only appearing on either block / writer assigns. Any reader methods won't trigger them now. You'll still get warnings, but I hope that you've gone from 4000+ to maybe 20-30 now across your suite. Can you try `2.17.1` and let me know if that's made it better (Without your monkeypatch)

If you feel inclined to do so you can try `3.0.beta`. It is reasonably stable and has about 70% of the work I wanted to implement for the 3.0 cut. Naturally as a major release there are a few changes (Namely all config settings are now removed). But I'll let you decide what's best at the moment for your use case.

@mnohai-mdsol - The reason for so many warnings was that any call to `merge_args` (Which any DSL method or DSL meta-programmed method will do), triggered the check to the `use_implicit_waits` reader config, and this in turned triggered a warning - Which probably was an oversight on my behalf. Feel free to chip in and review more if / when you have time.

:point_up: same goes for yourself Leo if you're interested in picking up any of the TODO's.

I'm going to close this as fixed in both `2.17.1` AND `3.0.beta`
